purepyvmware/datastores.py

```diff
@@ -0,0 +1,329 @@
+
+        Datastores.rescan_esxi_storage(connected_esxi_hosts)
+
+        return datastore
+
+
+class VvolDatastores(Datastores):
+    """Maintenance of VMware vVol datastores associated with Pure Storage FlashArrays."""
+
+    def __init__(self, sms_instance, vsphere_content, flasharray):
+        super(VvolDatastores, self).__init__(vsphere_content, flasharray)
+        self.sms_instance = sms_instance
+
+    def create_vvol_datastore(self, cluster_name, ds_name, flasharray, pe_name=None):
+        """Create a new vVol datastore on a Pure Storage FlashArray and connect it to the requested vSphere cluster.
+
+        Args:
+            cluster_name (str): Name of the cluster to create datastore in.
+            ds_name (str): Requested name of the datastore.
+            flasharray (purestorage.FlashArray): Existing session to FlashArray.
+            pe_name (optional, str): Name of the protocol endpoint to be used on a FlashArray.
+
+        Returns:
+            vvol_ds (vim.Datastore): Newly created vVol datastore object.
+        """
+        sc_id = None
+
+        if not pe_name:
+            pe_name = 'pure-protocol-endpoint'
+
+        hgroup, connected_esxi_hosts = self.verify_vsphere_cluster(cluster_name, flasharray)
+        existing_protocol_endpoints = flasharray.list_volumes(protocol_endpoint=True)
+
+        if connected_esxi_hosts:
+            if not existing_protocol_endpoints:
+                fa_protocol_endpoint = flasharray.create_conglomerate_volume(pe_name)
+                flasharray.connect_hgroup(hgroup, fa_protocol_endpoint['name'])
+            else:
+                for endpoint in existing_protocol_endpoints:
+                    if endpoint['name'] == pe_name:
+                        raise ValueError(f'\'{pe_name}\' already exists. Cancelling creation of vVol datastore.')
+                    else:
+                        fa_protocol_endpoint = flasharray.create_conglomerate_volume(pe_name)
+                        flasharray.connect_hgroup(hgroup, fa_protocol_endpoint['name'])
+
+        self.rescan_esxi_storage(connected_esxi_hosts)
+        storage_manager = self.sms_instance.QueryStorageManager()
+        storage_containers = storage_manager.QueryStorageContainer().storageContainer
+
+        for container in storage_containers:
+            # Example of arrayId output:
+            # (str) ['com.purestorage:3b7b308d-98f9-425e-87a1-3e57ada49658']
+            if container.arrayId[0].split(':')[1] == flasharray.get()['id']:
+                sc_id = container.uuid
+
+        for esxi_host in connected_esxi_hosts:
+            host_dssystem = esxi_host.configManager.datastoreSystem
+            vvol_spec = vim.HostDatastoreSystemVvolDatastoreSpec(name=ds_name, scId=sc_id)
+            vvol_datastore = host_dssystem.CreateVvolDatastore(vvol_spec)
+
+        self.rescan_esxi_storage(connected_esxi_hosts)
+
+        return vvol_datastore
+
+    def get_pure_provider(self, storage_manager, registration_url):
```

Comment:
could be static

Reply:
This may be something that we want to have overridable in the future, so for now I am going to keep it as an instance_method. 

-jhop